contracts/concentrated/PrivatePool.sol

```diff
@@ -0,0 +1,440 @@
+
+        return _D(xAdjustedWad, yAdjustedWad);
+    }
+
+    /// @notice Calculates amount of tokens received on swap
+    /// @dev Returns zero if pool balances exceeded on swap or invalid inputs
+    /// @param tokenIndexFrom The index of the token in
+    /// @param tokenIndexTo The index of the token out
+    /// @param dx The amount of token in in token decimals
+    function calculateSwap(
+        uint8 tokenIndexFrom,
+        uint8 tokenIndexTo,
+        uint256 dx
+    ) external view returns (uint256 dy_) {
+        (dy_, , ) = _calculateSwap(tokenIndexFrom, tokenIndexTo, dx);
+    }
+
+    /// @notice D liquidity param given pool token balances
+    /// @param xWad Balance of x tokens in wad
+    /// @param yWad Balance of y tokens in wad
+    function _D(uint256 xWad, uint256 yWad) internal view returns (uint256) {
+        return Math.mulDiv(xWad, P, wad) + yWad;
+    }
+
+    /// @notice Amount of token in wad
+    /// @param dx Amount of token in token decimals
+    /// @param isToken0 Whether token is token0
+    function _amountWad(uint256 dx, bool isToken0) internal view returns (uint256) {
+        uint256 factor = isToken0 ? 10**(token0Decimals) : 10**(token1Decimals);
+        return Math.mulDiv(dx, wad, factor);
+    }
+
+    /// @notice Amount of token in token decimals
+    /// @param amount Amount of token in wad
+    /// @param isToken0 Whether token is token0
+    function _amountDecimals(uint256 amount, bool isToken0) internal view returns (uint256) {
+        uint256 factor = isToken0 ? 10**(token0Decimals) : 10**(token1Decimals);
+        return Math.mulDiv(amount, factor, wad);
+    }
+
+    /// @notice Calculates amount of tokens received on swap
+    /// @dev Returns zero if pool balances exceeded on swap or invalid inputs
+    /// @param tokenIndexFrom The index of the token in
+    /// @param tokenIndexTo The index of the token out
+    /// @param dx The amount of token in in token decimals
+    // TODO: test with protocol fees
+    function _calculateSwap(
+        uint8 tokenIndexFrom,
+        uint8 tokenIndexTo,
+        uint256 dx
+    )
+        internal
+        view
+        returns (
+            uint256 dy_,
+            uint256 dyFee_,
+            uint256 dyAdminFee_
+        )
+    {
+        if (tokenIndexFrom > 1 || tokenIndexTo > 1 || tokenIndexFrom == tokenIndexTo) return (0, 0, 0);
+
+        // get current token balances in pool adjusted for protocol fees
+        uint256 xWad = _amountWad(IERC20(token0).balanceOf(address(this)) - protocolFees0, true);
+        uint256 yWad = _amountWad(IERC20(token1).balanceOf(address(this)) - protocolFees1, false);
+
+        // convert to an amount in wad
+        uint256 amountInWad = _amountWad(dx, tokenIndexFrom == 0);
+
+        // calculate swap amount out wad
+        // @dev obeys P * x + y = D
+        uint256 amountOutWad;
+        if (tokenIndexFrom == 0) {
+            // get D balance before swap
```

Comment:
Realized that this could be changed to `amountOutWad = Math.mulDiv(amountInWad, P, wad)` (and the check that LP balance covers that).

Reply:
Agree here 💯